### ISSUE TITLE
byuu's historical patching tools

### ISSUE THREAD

**Issue by Rus3545** [Screwtapello/beat#1]
Thank. But **bpm** is still not there.
Here is version 03.
[Link](https://github.com/arizvisa/byuu.org-beat)
`beat_v03.tar\beat_v03.tar\nall\beat\multi.hpp`
But there bpm is already broken.
I don’t remember if it worked at 02, but at 01 it worked for sure.

As i understand it in the new version 1.x it _will be called **bpa**_(incorrectly), but the functionality is not yet involved.

P.S.
[Sample](http://shedevr.org.ru/zelda64rus/translations/mm_3D_rus/LoZMM3D_(PAL_rev1)_RUS_Patch_v1.1.rar)
I wanted to add elf and sh, but met with a terrible fate.

**Comment by Screwtapello** [comment]
I set up this repo because byuu asked me to, and he specifically asked me not to include older, historical patching tools because the new version of beat was created from scratch.

Although I have byuu's old single-file patch tools, I didn't bother to archive the multi-file patch tools. I found an original-looking tarball for beat v03, but I haven't yet managed to find v02 or v01.

If you can run the original Windows version of beat v01 in Wine, perhaps you can apply the patches manually to recreate the target files, and then create new patches with beat v03, so they'll be compatible with a version of beat for which source-code survives.

Better yet, run the script on Windows to create a translated ROM, and then use a tool like [Flips](https://github.com/Alcaro/Flips) to make a BPS patch from the old to the new version of the ROM, instead of extracting files, patching them, then re-injecting them into the ROM. Unlike the BPM format, which was never finalised and probably shouldn't be used, the BPS format is stable and has multiple, independent, compatible implementations.

**Comment by Rus3545** [comment]
I apologize for the difficulties of translation and continuation of the discussion.
But you don't understand me.
Versions 01 and 03 create absolutely identical patches, but version 03 will never be able to apply its patch, since Apply Patch = Crash, i don’t remember the  regression only it or dragged on already from 02 and a radical remake of CLI commands.
I found version 01 on the network only for Windows and my searches were apparently empty.
But I need exactly the version for Linux, because there was an opportunity to rewrite bat to sh, and therefore asked for beat_v01.tar.xz.
Ok. Eureka. I'll try to fix [it](http://helmet.kafuka.org/byuubackup/viewtopic.php@f=10&t=1919&start=150.html) for 03...


P.S.
The sizes of the bps and bpm patches are incommensurable when applied to an encrypted image and to a decrypted group of files, and there is no binding to the crc image, the main thing is that the files are original inside. So bpm is better here.
P.S. 2
Flips cannot patch large images, for example PSX, but xdelta3 and beat_v01 in one of the modes do it. 
But the size of the patches... Or do a lot of **bps** patches.
It was very surprising when I first saw that comparing folders gives one small patch, which also compresses perfectly. 40mb modified files -> ~ 3mb **bpm**, which will be compressed by another 0.5mb when packaging for Oot 3D patch. That was a very good idea .
It was a very good idea to patching for Wii, GCN and other rebuilded things.

+Another historical patcher: [Link](https://traf.romhack.org/?p=outils&oid=182).

**Comment by Screwtapello** [comment]
> Versions 01 and 03 create absolutely identical patches, but version 03 will never be able to apply its patch, since Apply Patch = Crash

Ah, that's a shame.

I've looked and asked people, but I haven't managed to find a copy of `beat_v01.tar.xz`. I also checked the bsnes repo (since `nall/beat/multi.hpp` is part of the `nall` library that bsnes also uses), but the oldest copy of that file seems to correspond to v03.

If you ever do manage to find a copy of v01 or v02, I'd love to get a copy for myself!

> The sizes of the bps and bpm patches are incommensurable when applied to an encrypted image and to a decrypted group of files

Ah, yeah, encryption (or compression) screws up delta-patchers quite badly. I was under the impression that 3DS games were usually decrypted as part of the dumping process, but I guess that's not always the case.

> Flips cannot patch large images, for example PSX, but xdelta3 and beat_v01 in one of the modes do it.

xdelta3 is great, although the file-format is quite complex.

beat's "linear" patching mode should also be fast, but usually generates terrible, inefficient patches.

The modern version of beat should be much smarter and more efficient than the old version, but I don't think it's been tested beyond a ~300MB Sega CD image.

I haven't tested Flips on CD images myself, but [when I recently tested it](https://zork.net/~st/jottings/delta-compression-tests-2019.html) it was a bit better than modern beat, so I hoped it would be enough to handle an uncompressed, unencrypted 3DS image. If not, that's a shame. :/

**Comment by Rus3545** [comment]
Thanks. Still, i figured out this issue.
I don’t like only the CLI Application window on the background of the program on Windows.
For Linux, it does not appear and everything is beautiful there.
Everything else suits me now.
My [fork beat_03+](https://github.com/Rus3545/byuu.org-beat).
03+ = 03 -> 03_fix+01_CLI.

P.S.
> xdelta3 is great, although the file-format is quite complex.

Yep. But sometimes it can patch files with different crc without using this key or just try and leave a piece of garbage if something is not what was expected at the input. Everyone is wrong...

**Comment by Rus3545** [comment]
> I haven't tested Flips on CD images myself, but when I recently tested it

Thank. Informative...
About compression (bonus).
I used a combination of xdelta3-9 + ecm + scripts to pack absolutely all versions (US, JPN, and all_Eur_for_PE_II) of Parasite_Eve_I & Parasite_Eve_II for PSX into a ~ 1Gb 7z archive and unpack them back. Donors basis are US versions of both titles, maybe even only the first discs, i don’t remember the details.
It was the same amazing experiment as with bpm when we had to solve an unusual problem for patching: [Link](http://zelda64rus.ucoz.ru/forum/33-670-24188-16-1496084734).

Which eventually evolved into a standard script form for these patches.
There are still *.3ps patches that 3dstool makes, but they were used just like crutches, so i don’t even know what potential 3dstool has as a patcher, but the Yaz0 codec is very fast there, in addition to the main functionality ...

And again about bpm.
At the first test, beat_02 or beat_03 in 2013 did not pay attention to bpm, most likely precisely because [it did not work](http://zelda64rus.ucoz.ru/forum/14-320-19789-16-1382472240). As far as I see the format is quite finished, these are the same bps containing in the header the locations of the modified files in the folder structure and glued into one common file. 

Well, i also tested various patcher once too: [Link](http://zelda64rus.ucoz.ru/forum/32-677-24353-16-1508130601).
